File: dd-java-agent/instrumentation/java-concurrent/src/test/java/AsyncChild.java
import datadog.trace.api.Trace;
import java.util.concurrent.Callable;
import java.util.concurrent.atomic.AtomicBoolean;

public class AsyncChild implements Runnable, Callable {
  private final AtomicBoolean blockThread;
  private final boolean doTraceableWork;

  public AsyncChild() {
    this(true, false);
  }

  public AsyncChild(final boolean doTraceableWork, final boolean blockThread) {
    this.doTraceableWork = doTraceableWork;
    this.blockThread = new AtomicBoolean(blockThread);
  }

  public void unblock() {
    blockThread.set(false);
  }

  @Override
  public void run() {
    runImpl();
  }

  @Override
  public Object call() throws Exception {
    runImpl();
    return null;
  }

  private void runImpl() {
    if (doTraceableWork) {
      asyncChild();
    }
    while (blockThread.get()) {
      // busy-wait to block thread
    }
  }

  @Trace(operationName = "asyncChild")
  private void asyncChild() {}
}

File: instrumentation/scala-fork-join-2.8/javaagent/src/test/scala/io/opentelemetry/javaagent/instrumentation/scalaexecutors/AsyncChild.scala
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.javaagent.instrumentation.scalaexecutors

import io.opentelemetry.api.GlobalOpenTelemetry
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.javaagent.instrumentation.executors.TestTask

import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicBoolean
import scala.concurrent.forkjoin.ForkJoinTask

final class AsyncChild(doTraceableWork: Boolean, blockThread: Boolean)
    extends ForkJoinTask[AnyRef]
    with TestTask {
  private val tracer: Tracer = GlobalOpenTelemetry.getTracer("test")
  private val blockThreadAtomic: AtomicBoolean = new AtomicBoolean(blockThread)
  private val latch: CountDownLatch = new CountDownLatch(1)

  override def getRawResult: AnyRef = null

  override def setRawResult(value: AnyRef): Unit = {}

  override def exec(): Boolean = {
    runImpl()
    true
  }

  override def unblock(): Unit = {
    blockThreadAtomic.set(false)
  }

  override def run(): Unit = {
    runImpl()
  }

  override def call(): AnyRef = {
    runImpl()
    null
  }

  override def waitForCompletion(): Unit = {
    try {
      latch.await()
    } catch {
      case e: InterruptedException =>
        Thread.currentThread().interrupt()
        throw new AssertionError(e)
    }
  }

  private def runImpl(): Unit = {
    while (blockThreadAtomic.get()) {
      // busy-wait to block thread
    }
    if (doTraceableWork) {
      asyncChild()
    }
    latch.countDown()
  }

  private def asyncChild(): Unit = {
    tracer.spanBuilder("asyncChild").startSpan().end()
  }
}
